Add chapter on testing interceptors (#150)
infinum/frontend-handbook: Angular/Angular guidelines and best practices/Testing.md

}
```

## Testing a interceptor

> If interceptor depends on some services we should mock and inject those services into `TestBed` as we did in example with regular service. 

Reviewer: ```suggestion
If an interceptor depends on some services, we should mock and inject those services via `TestBed`, just like in the example with a regular service.
```

> Let's say that this interceptor injects `AuthenticationService` which handles login logic and state. 

Reviewer: ```suggestion
As an example, let's say we have an interceptor that injects `AuthenticationService` which handles login logic and state.
```

> Mock implementation of `AuthenticationTestingService` for our purpose could look something like:

Reviewer: ```suggestion
Mock implementation of `AuthenticationTestingService` for our purpose could look something like this:
```

```ts
export class AuthenticationTestingService implements ExtractPublic<AuthenticationService>{
  public getAccessToken(): string{
    return 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTY0MzEwNTA2MywiZXhwIjoxNjQzMTA4NjYzfQ.6JOJdydKd-jzZMDWiqr2mUvC78DIwJNd0ye-OZOymGg' // JTW token generated via https://token.dev/
  }
}
```

> As with regular services we create test double for `AuthenticationService`, and pass it to `TestBed` configuration, also we need to register our interceptor via the `HTTP_INTERCETPORS` multi provider token. 

Reviewer: ```suggestion
As with regular services, we create a test double for `AuthenticationService`, and pass it to the `TestBed` configuration. Additionally, we need to register our interceptor via the `HTTP_INTERCETPORS` multi provider token.
```

```ts
describe('AuthorizationInterceptor', () => {
  let httpClient: HttpClient;
  let httpMock: HttpTestingController;
  let authenticationService: AuthenticationService;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [HttpClientTestingModule],
      providers: [
        { provide: AuthenticationService,
          useClass: AuthenticationTestingService // Provide mock dependency
        }, 
        {
          provide: HTTP_INTERCEPTORS, // Register interceptor
          useClass: AuthorizationInterceptor,
          multi: true
        }
      ],
    });

      httpClient = TestBed.inject(HttpClient);
      httpMock = TestBed.inject(HttpTestingController); 
      authenticationService = TestBed.inject(AuthenticationService);
  });

  afterEach(() => {
    httpMock.verify();
  });
  ```

>   When testing services that make http calls it is recommended that afterEach test we check wether there are any outstanding API calls which weren't resolved, we achieve this with calling `verify()` method on our testing instance of `HttpTestingController`. 

Reviewer: This explanation is redundant because we already mention that verify should be called in afterEach, in `Mocking HTTP requests in tests` sub-chapter. You can remove this explanation.

>   Below are relevant tests regarding interceptor functionality.

Reviewer: ```suggestion
Below are relevant tests regarding the interceptor's functionality.
```

  ```ts

>   it('should attach the interceptor', () => {

Reviewer: Fix indentation

    const interceptors: Array<HttpInterceptor> = TestBed.inject(HTTP_INTERCEPTORS);
      expect(interceptors.some((i) => i instanceof AuthorizationInterceptor)).toBe(true);
  });

  it('should set authorization header to request', () => {
      const url = 'https://my-api.com/api/some-route';
    
      httpClient.get(url).subscribe();

      const mockRequest = httpMock.expectOne(url);

      const authHeader = mockRequest.request.headers.get('Authorization');

      expect(authHeader).toBeTruthy();
      expect(authHeader).toBe(`Bearer ${authenticationService.getAccessToken()}`);

      mockRequest.flush(null);
  });
})
```

> To learn more about testing services that make http requests read official Angular documentation regarding [Testing HTTP requests](https://angular.io/guide/http#testing-http-requests)

Reviewer: This sentence would make more sense in `Testing HTTP requests` sub-chapter of the handbook because the linked documentation does not mention anything related to testing interceptors, so it does not make too much sense to place this comment in a sub-chapter that is only about testing interceptors. Can you please move this sentence to the mentioned sub-chapter?

Also, update the wording:

```suggestions
To learn more, read the official Angular documentation chapter about [Testing HTTP requests](https://angular.io/guide/http#testing-http-requests)
```
